c/tskit/trees.c

```diff
@@ -2076,19 +2076,567 @@ typedef struct {
+    tsk_size_t mutation_index, allele, alt_allele_length;
+    /* The allele table */
+    tsk_size_t max_alleles = site->mutations_length + 1;
+    const char **alleles = tsk_malloc(max_alleles * sizeof(*alleles));
+    tsk_size_t *allele_lengths = tsk_calloc(max_alleles, sizeof(*allele_lengths));
+    const char *alt_allele;
+    tsk_bit_array_t state_row;
+    tsk_bit_array_t allele_samples_row;
+    tsk_bit_array_t alt_allele_samples_row;
+    tsk_size_t num_alleles = 1;
+
+    if (alleles == NULL || allele_lengths == NULL) {
+        ret = TSK_ERR_NO_MEMORY;
+        goto out;
+    }
+
+    tsk_bug_assert(state != NULL);
+    alleles[0] = site->ancestral_state;
+    allele_lengths[0] = site->ancestral_state_length;
+
+    for (mutation_index = 0; mutation_index < site->mutations_length; mutation_index++) {
+        mutation = site->mutations[mutation_index];
+        /* Compute the allele index for this derived state value. */
+        for (allele = 0; allele < num_alleles; allele++) {
+            if (mutation.derived_state_length == allele_lengths[allele]
+                && tsk_memcmp(
+                       mutation.derived_state, alleles[allele], allele_lengths[allele])
+                       == 0) {
+                break;
+            }
+        }
+        if (allele == num_alleles) {
+            tsk_bug_assert(allele < max_alleles);
+            alleles[allele] = mutation.derived_state;
+            allele_lengths[allele] = mutation.derived_state_length;
+            num_alleles++;
+        }
+
+        /* Add the mutation's samples to this allele */
+        tsk_bit_array_get_row(out_allele_samples, allele, &allele_samples_row);
+        tsk_bit_array_get_row(state, mutation_index, &state_row);
+        tsk_bit_array_add(&allele_samples_row, &state_row);
+
+        /* Get the index for the alternate allele that we must subtract from */
+        alt_allele = site->ancestral_state;
+        alt_allele_length = site->ancestral_state_length;
+        if (mutation.parent != TSK_NULL) {
+            parent_mut = site->mutations[mutation.parent - site->mutations[0].id];
+            alt_allele = parent_mut.derived_state;
+            alt_allele_length = parent_mut.derived_state_length;
+        }
+        for (allele = 0; allele < num_alleles; allele++) {
+            if (alt_allele_length == allele_lengths[allele]
+                && tsk_memcmp(alt_allele, alleles[allele], allele_lengths[allele])
+                       == 0) {
+                break;
+            }
+        }
+        tsk_bug_assert(allele < num_alleles);
+
+        tsk_bit_array_get_row(out_allele_samples, allele, &alt_allele_samples_row);
+        tsk_bit_array_subtract(&alt_allele_samples_row, &allele_samples_row);
+    }
+    *out_num_alleles = num_alleles;
+out:
+    tsk_safe_free(alleles);
+    tsk_safe_free(allele_lengths);
+    return ret;
+}
+
+static int
+get_mutation_samples(const tsk_tree_t *self, const tsk_size_t *site_offsets,
```

Comment:
I think it'd be tough to use this function in isolation, I would consider it to be an internal function until decided otherwise. In that vain, I've changed this parameter to `tree` instead of `self`.